### PR TITLE
Homework - Unique_ptr

### PR DESCRIPTION
**Pull request by Pawellos**
Implementation own Unique_ptr (constructor, move constructor, destructor, operatos, methods: get, reset, release, removed copy constructor)

### CODE REVIEW

File: homework/unique_ptr/UniquePointer.hpp [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+template <class T>
+UniquePointer<T>::UniquePointer() {}
+
+template <class T>
+UniquePointer<T>::UniquePointer(T* ptr) : ptr_(ptr) {}
+
+template <class T>
+UniquePointer<T>::~UniquePointer() {
+    if (ptr_) {
```
Comment (MaciejSalamonski) [inline homework/unique_ptr/UniquePointer.hpp:38]: I think that You don't have to check if ptr isn't null pointer in destructor

https://stackoverflow.com/questions/50569942/check-pointer-is-not-null-in-destructor

Delete or delete[] are defined for nullptr. They do nothing

**Review by MaciejSalamonski — APPROVED** [review]
Hi nice code! Have a great day ;~}

File: homework/unique_ptr/UniquePointer.hpp (hunk 1 above)
Comment (Pawellos) [reply homework/unique_ptr/UniquePointer.hpp:38]: Ok, thanks. I was not sure :)

File: homework/unique_ptr/UniquePointer.hpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    UniquePointer();
+    ~UniquePointer();
+    UniquePointer(const UniquePointer&) = delete;
+    UniquePointer(UniquePointer&& anotherPtr);
+
+    T* get();
+    T* release();
+    void reset(T* ptr);
+
+    T operator*() const;
```
Comment (ziobron) [inline homework/unique_ptr/UniquePointer.hpp:21]:
```suggestion
    T& operator*() const;
```
This operator should return a reference, not a copy

File: homework/unique_ptr/UniquePointer.hpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    T* operator->() const;
+    UniquePointer<T>& operator=(UniquePointer<T>&& anotherPtr);
+    UniquePointer<T>& operator=(const UniquePointer<T>&) = delete;
+
+private:
+    T* ptr_;
+};
+
+template <class T>
+UniquePointer<T>::UniquePointer() {}
```
Comment (ziobron) [inline homework/unique_ptr/UniquePointer.hpp:31]: default constructor has uninitialized pointer, so it causes undefined behavior ❗️

File: homework/unique_ptr/test.cpp [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+struct TestUniquePtr : ::testing::Test {
+    TestUniquePtr() : uPtr(new int{testValue1}) {}
+    UniquePointer<int> uPtr;
+};
+
+TEST_F(TestUniquePtr, TestOfConstructor) {
+    ASSERT_EQ(testValue1, *uPtr);
+
+    UniquePointer<int> ptr{new int{33}};
+    ASSERT_EQ(33, *ptr);
```
Comment (ziobron) [inline homework/unique_ptr/test.cpp:17]: avoid magic values. Name them.

File: homework/unique_ptr/test.cpp [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    UniquePointer<int> ptr(new int(15));
+    ptr.reset(new int(45));
+    ASSERT_EQ(45, *ptr);
+}
+
+TEST_F(TestUniquePtr, TestOfReleaseMethod) {
+    UniquePointer<int> ptr(new int(35));
+    ptr.release();
+    ASSERT_EQ(ptr.get(), nullptr);
+}
```
Comment (ziobron) [inline homework/unique_ptr/test.cpp:42]: there is a memory leak in this test case. You should assign a result of release() function to a pointer and delete it.

File: homework/unique_ptr/test.cpp [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+TEST_F(TestUniquePtr, TestOfNullPtrException) {
+    UniquePointer<int> ptr{nullptr};
+    ASSERT_THROW(*ptr, NullPtrException);
+}
+
+TEST_F(TestUniquePtr, TestOfMoveConstructor) {
+    UniquePointer<int> ptr(std::move(uPtr));
+    ASSERT_EQ(20, *ptr);
```
Comment (ziobron) [inline homework/unique_ptr/test.cpp:51]: What is 20? Why 20? Magic value.
What about uPtr after move? Also check if it's in a valid state.

File: homework/unique_ptr/test.cpp [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+TEST_F(TestUniquePtr, TestOfMoveAssignmentOperator) {
+    UniquePointer<int> ptr{new int{14}};
+    UniquePointer<int> ptr2 = std::move(ptr);
+    ASSERT_EQ(14, *ptr2);
```
Comment (ziobron) [inline homework/unique_ptr/test.cpp:57]: what about ptr? you should also test it.

**Review by ziobron — CHANGES_REQUESTED** [review]

**Comment by ziobron** [comment]
And it will be useful to have a CI job with valgrind / asan to prove that there are no memory leaks.

**Review by WBigaj — COMMENTED** [review]
I have checked your code and have nothing to add that have not been mentioned by other reviewers.

**Comment by ziobron** [comment]
Ok, but please move homework/unique_ptr/.github/workflows/unique.yml to .github/workflows/unique.yml. It will trigger CI and I will be able to check results.

**Comment by Pawellos** [comment]
> Ok, but please move homework/unique_ptr/.github/workflows/unique.yml to .github/workflows/unique.yml. It will trigger CI and I will be able to check results.

Ok, I think it's done  :)

**Comment by ziobron** [comment]
Invalid workflow file
You have an error in your yaml syntax on line 7

Probably too many spaces in indentation. 4 instead of 2. Check it.

File: .github/workflows/unique.yml [hunk 8]
```diff
@@ -0,0 +1,38 @@
+name: C++ with CMake CI
+
+on: [push, pull_request]
+
+jobs:
+  UniquePtr:
+      runs-on: ubuntu-20.04
```
Comment (ziobron) [inline .github/workflows/unique.yml:7]:
```suggestion
    runs-on: ubuntu-20.04
```

File: .github/workflows/unique.yml [hunk 9]
```diff
@@ -34,5 +34,5 @@ jobs:
     - name: Run valgrind 
       working-directory: homework/unique_ptr/build
       run: |
-        valgrind ./UniquePtr-ut
-        valgrind ./UniquePtr
+        valgrind ./test
```
Comment (ziobron) [inline .github/workflows/unique.yml:37]: no such file or directory ./test. How is your UT binary named?

Comment (Pawellos) [reply .github/workflows/unique.yml:37]: I am lost with this. I think it is test.cpp

Comment (ziobron) [reply .github/workflows/unique.yml:37]: You can get this from CMakeLists.txt

```cmake
project(unique_ptr)
 enable_testing()

 add_executable(${PROJECT_NAME} main.cpp)
 add_executable(${PROJECT_NAME}-ut test.cpp)
```

It is `${PROJECT_NAME}-ut`, so it's `unique_ptr-ut`

Comment (Pawellos) [reply .github/workflows/unique.yml]: Thanks ! I was messing with this a while and i was not sure what is wrong. Anyway thanks for help :)

**Comment by ziobron** [comment]
{ "requirement_ids": [3784, 1119] }
36 XP za implementację unique_ptra (3784)
3 XP za punktualność

**Pull request closed without merging**